File: python/56_Merge_Intervals.py
'''
Time: O(nlogn)  -- sorting
'''

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key = lambda i : i[0])  # sort by 1st value of each interval
        output = [intervals[0]]  # initialize with the 1st interval

        for start, end in intervals[1:]:  # current start/ends
            lastEnd = output[-1][1]

            if start <= lastEnd:  # overlapping
                output[-1][1] = max(lastEnd, end)
            else:
                output.append([start, end])

        return output

File: javascript/56_Merge_Intervals.js
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function(intervals) {
    intervals.sort((a, b) => a[0] - b[0]);
    let output = [intervals[0]];

    for (let [start, end] of intervals.slice(1)) {
        const lastEnd = output[output.length-1][1];
        if (start <= lastEnd) output[output.length-1][1] = Math.max(lastEnd, end);
        else output.push([start, end]);
    }
    return output;
};